### DISCUSSION TITLE
"Inintialize" function for a group of qbit?

### DISCUSSION THREAD

**Discussion by RechercheStochastique** [anyonlabs/Snowflake.jl#27] [Q&A] [answered]
Is there a way to setup the state of several qbits with initial state? Something like the Initialise() method of qiskit? For example in qiskit:

circuit = QuantumCircuit(2)
circuit.initialize([0, 1/np.sqrt(2), -1.j/np.sqrt(2), 0], circuit.qubits)

**Comment by npoirie1** [comment, marked as answer]
Hello Mr. Gagnon,
At the start of the circuit, the qubits are in the ground state |00....0>. There are currently no functions to initialize the qubits in a different state. For the moment, you need to apply the gates which are available in Snowflake to reach that state. A function that determines which gates must be applied to reach a desired state may be implemented in the near future.

**Reply by RechercheStochastique** [reply]
OK.
Then, do you know a simple way to do this? Can you point me to a document explaining how to do that with Snowflake functions? For qiskit there was a "phase" gate where I could define the phase shit by an arbitrary number. Anything equivalent with Snowflake? 

Est-ce que je peut poser me question en français?

**Reply by npoirie1** [reply]
The phase gate along with other rotation gates are available in the feat/add-parameterized-quantum-gates branch. This branch will be merged into the main branch shortly. If you want to use the feature branch, you can git clone Snowflake and git checkout the feature branch from the Snowflake.jl/ folder.

From the Snowflake.jl/ folder, you can then call:
julia
julia> ]
(@v1.7) pkg> activate .
Press backspace
julia> using Snowflake
julia> phase_shift(1, 0)

To have a look at the Snowflake documentation locally, you can call:
julia> ]
(@v1.7) pkg> activate docs
(docs) pkg> instantiate
(docs) pkg> dev .
Press backspace
julia> include("docs/make.jl")

This will generate html files in the docs/build/ directory. The library.html file contains a list of all the quantum gates.

After the next release, the Snowflake documentation will be available online. This will make finding Snowflake functions much easier.

**Reply by npoirie1** [reply]
To determine the gates which are needed to generate a certain quantum state from the ground state, you can follow the approach described by Shende, Bullock, and Markov: https://arxiv.org/abs/quant-ph/0406176. The idea is to start from the desired quantum state and determine which quantum gates must be applied to reach the ground state. This is done by disentangling qubits one-by-one using Theorem 9 of Section 4. The result will be a circuit with multiplexors. The multiplexors can be decomposed into rotation and control_x gates as shown in Section 3.1. The circuit is then reversed such that the desired state can be generated from the ground state.

For example, the state [0, 1/sqrt(2), -im/sqrt(2), 0] can be generated as follows:

julia> c = QuantumCircuit(qubit_count=2, bit_count=0);
julia> push_gate!(c, rotation_x(1,π/2));
julia> push_gate!(c, control_x(1,2));
julia> push_gate!(c, rotation_y(2,π/2));
julia> push_gate!(c, control_x(1,2));
julia> push_gate!(c, rotation_y(2,π/2));
julia> simulate(c)
4-element Ket:
1.1102230246251565e-16 + 0.0im
0.7071067811865476 - 4.329780281177467e-17im
4.329780281177465e-17 - 0.7071067811865475im
-4.329780281177466e-17 + 0.0im

**Reply by npoirie1** [reply]
Adding a function which generates the gates needed to reach a certain quantum state is currently not the priority. However, we more than welcome contributions if someone wants to implement that feature.

**Reply by npoirie1** [reply]
Vous pouvez me poser vos questions en français aussi.

**Reply by RechercheStochastique** [reply]
Merci beaucoup. 
C'est comme ça que je comptait aborder le problème mais je me disais que quelqu'un avait justement fait tous les calculs.
J'ai effectivement 2^K valeurs à générer pour les K qubits que je veux utiliser.
